### PR TITLE
Required Model Section Check

### PR DESCRIPTION
**Pull request by mousedownmike**
For #333 this PR checks for the existence of `r`,`p`,`e` and `m` model sections, returning an error when loading if any of those is not present.

This also has a minor typo correction in a comment and a move of the empty string check in `model.AddDef()`.

### CODE REVIEW

File: persist/adapter_test.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package persist
+
+import (
+	"github.com/casbin/casbin/v2/model"
+	"testing"
```
Comment (hsluoyz) [inline persist/adapter_test.go]: Sort the imports, see other files.

File: persist/adapter_test.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package persist
```
Comment (hsluoyz) [inline persist/adapter_test.go]: License header missing.

File: model/model.go [hunk 3]
```diff
@@ -140,16 +144,22 @@ func (model Model) LoadModelFromText(text string) error {
...
-	loadSection(model, cfg, "m")
-
-	loadSection(model, cfg, "g")
-
+	for s := range sectionNameMap {
+		loadSection(model, cfg, s)
+	}
+	for _, rs := range requiredSections {
+		if !model.hasSection(rs) {
+			return errors.New("model does not contain all required sections ")
```
Comment (hsluoyz) [inline model/model.go]: Show the missing section's name, remove trailing space.

File: model/model_test.go [hunk 4]
```diff
@@ -1,9 +1,102 @@
+// Copyright 2019 The casbin Authors. All Rights Reserved.
```
Comment (hsluoyz) [inline model/model_test.go]: Put them into a new `config_test.go` file.

Comment (mousedownmike) [reply model/model_test.go]: I think you're referring to the `MockConfig` I've defined here.  I was using that to isolate the tests without using a complete `config.ConfigInterface` implementation.  If that gets moved to `config_test.go` it no longer becomes usable in this test.  I haven't found a consistent pattern for where people place their mocks in Go.  Would moving the Mock definitions to the bottom of the file be better?

**Comment by mousedownmike** [comment]
Apologies for the `adapter_test.go` file getting wrapped up in this PR.  I've improved the error message contents when sections are missing.  I need clarification on your `config_test.go` comment, maybe just move the mocks to the bottom of the file?

I merged in an additional test I created for `model.AddDef`, I was using it to verify my value check change but it didn't get merged in earlier.

**Comment by mousedownmike** [comment]
@hsluoyz  let me know if you need to make any additional changes here, I'd be happy to.

**Comment by hsluoyz** [comment]
Hi @mousedownmike please rework your commits. Any "Merge", "Need to revert" commits should be replaced.

image

**Comment by hsluoyz** [comment]
Just provide the final-version, good-looking, easy-to-review commits

**Comment by mousedownmike** [comment]
That makes sense but I'm not sure how to do that with an existing PR in GitHub so I'm going to close this PR and open a new one with the commits squashed.

**Pull request closed without merging**